src/NumericalAlgorithms/Spectral/SwshCollocation.hpp

```diff
@@ -0,0 +1,245 @@
+struct DestroySharpGeometry {
+  void operator()(sharp_geom_info** to_delete) noexcept {
+    sharp_destroy_geom_info(*to_delete);
+    free(to_delete);
+  }
+};
+}  // namespace detail
+
+/// A container for reporting a single collocation point for libsharp compatible
+/// data structures
+struct LibsharpCollocationPoint {
+  size_t offset;
+  double theta;
+  double phi;
+};
+
+/// \brief A wrapper class for the spherical harmonic library collocation data
+///
+/// \details The currently chosen library for spin-weighted spherical harmonic
+/// transforms is libsharp. The `Collocation` class stores the
+/// libsharp `sharp_geom_info` object, which contains data about
+/// 1. The angular collocation points used in spin-weighted spherical harmonic
+/// transforms
+/// 2. The memory representation of double-type values at those collocation
+/// points
+///
+/// \tparam Representation the ComplexRepresentation, either
+/// `ComplexRepresentation::Interleaved` or
+/// `ComplexRepresentation::RealsThenImags` compatible with the generated
+/// `Collocation` - this is necessary because the stored libsharp type contains
+/// memory stride information.
+template <ComplexRepresentation Representation>
+class Collocation {
+ public:
+  class CollocationIterator {
+   public:
+    /// create a new iterator. defaults to the start of the supplied object
+    explicit CollocationIterator(
+        const gsl::not_null<const Collocation<Representation>*> collocation,
+        size_t start_index = 0) noexcept
+        : index_{start_index}, collocation_{collocation} {}
+
+    /// recovers the data at the collocation point using a
+    /// `LibsharpCollocationPoint`, which stores the vector `offset` of the
+    /// location of the `theta`, `phi` point in libsharp compatible data
+    LibsharpCollocationPoint operator*() const noexcept {
+      return LibsharpCollocationPoint{index_, collocation_->theta(index_),
+                                      collocation_->phi(index_)};
+    }
+    /// advance the iterator by one position (prefix)
+    CollocationIterator& operator++() noexcept {
+      ++index_;
+      return *this;
+    };
+    /// advance the iterator by one position (postfix)
+    // clang-tidy wants this to return a const iterator
+    CollocationIterator operator++(int) noexcept {  // NOLINT
+      return CollocationIterator(collocation_, index_++);
+    }
+
+    /// retreat the iterator by one position (prefix)
+    CollocationIterator& operator--() noexcept {
+      --index_;
+      return *this;
+    };
+    /// retreat the iterator by one position (prefix)
+    // clang-tidy wants this to return a const iterator
+    CollocationIterator operator--(int) noexcept {  // NOLINT
+      return CollocationIterator(collocation_, index_--);
+    }
+
+    // @{
+    /// (In)Equivalence checks both the object and index for the iterator
+    bool operator==(const CollocationIterator& rhs) noexcept;
```

Comment:
I think `operator==` can be a `const` member function (`const noexcept;` at the end :) )